core/plugins/stack/terraform/version.ts

```diff
@@ -0,0 +1,87 @@
+import { IntrospectFn } from "../../../types.ts";
+
+// Find the latest stable version here:
+// https://www.terraform.io/downloads
+const LATEST = "1.1.2";
+
+const WARN_USING_LATEST =
+  `Couldn't detect the Terraform version, using the latest available: ${LATEST}`;
+
+const ERR_UNDETECTABLE_TITLE =
+  "Couldn't detect which Terraform version this project uses.";
+const ERR_UNDETECTABLE_INSTRUCTIONS = `
+To fix this issue, consider one of the following suggestions:
+
+1. Add the 'required_version' to your .tf file
+
+See https://www.terraform.io/language/settings#specifying-a-required-terraform-version
+
+2. Create a .terraform-version file
+
+The .terraform-version file can be used by the tfenv to choose a specific Terraform version
+for a project.
+
+It's the easiest option, all you have to do is create a .terraform-version text
+file with a version inside, like "1.1.0".
+
+See https://github.com/tfutils/tfenv
+`;
+
+const terraformVerRegex =
+  /required_version.*(?<TerraformVersion>[0-9]+\.[0-9]+\.[0-9])/gm;
+
+function notEmpty<TValue>(value: TValue | null | undefined): value is TValue {
+  return value !== null && value !== undefined;
+}
+
+const tfenv: IntrospectFn<string | Error> = async (context) => {
+  for await (const file of context.files.each("**/.terraform-version")) {
+    return await context.files.readText(file.path);
+  }
+  return Error("Can't find terraform version at .terraform-version");
+};
+
+const tfFile: IntrospectFn<string | Error> = async (context) => {
+  for await (const file of context.files.each("**/*.tf")) {
+    const tfText = await context.files.readText(file.path);
+
+    const terraformVersion: string[] = Array.from(
+      tfText.matchAll(
+        terraformVerRegex,
+      ),
+      (match) => !match.groups ? null : match.groups.TerraformVersion,
+    ).filter(notEmpty);
+
+    if (terraformVersion.length > 0) {
+      return terraformVersion[0];
+    }
```

Comment:
If you want only the first match, you could remove the global modifier from the regex (the `g` at the end) and write something like:
```typescript
    const matchGroups = tfText.match(tfText)?.groups || {};
    if ("TerraformVersion" in matchGroups) {
      return matchGroups["TerraformVersion"];
    }
```